### PR TITLE
fix(pairing): fix issues with unverified sessions and lack of sessionToken

### PR DESCRIPTION
**Pull request by vladikoff**
~~Draft to fix pair edge cases with unverified accounts. Gonna add some tests and clean up~~

Fixes https://github.com/mozilla/fxa/issues/923

### CODE REVIEW

File: packages/fxa-content-server/app/scripts/views/pair/index.js [hunk 1]
```diff
@@ -26,17 +28,23 @@ class PairIndexView extends FormView {
...
-    if (! this.broker.get('browserSignedInAccount')) {
-      // if we are not logged into Sync then we offer to sign in
+    if (! account) {
+      // if we are not logged into Sync then we offer to connect
       return this.replaceCurrentPage('connect_another_device');
     }
 
+    if (! account.verified || ! account.sessionToken) {
+      // if account is not verified or missing sessionToken then offer to sign in or confirm
+      return window.location.href = this.getEscapedSyncUrl('signin', 'fxa:pair');
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/pair/index.js]: Can you do a `this.navigateAway` with the URL here so that metrics are flushed before redirecting?

**Comment by vladikoff** [comment]
@shane-tomlinson updated !

File: packages/fxa-content-server/app/scripts/views/pair/index.js [hunk 2]
```diff
@@ -21,22 +22,27 @@ class PairIndexView extends FormView {
...
-      // if we are not logged into Sync then we offer to sign in
+    if (! account) {
+      // if we are not logged into Sync then we offer to connect
       return this.replaceCurrentPage('connect_another_device');
     }
 
-    if (! this.broker.hasCapability('supportsPairing')) {
-      return this.replaceCurrentPage('pair/unsupported');
+    if (! account.verified || ! account.sessionToken) {
+      // if account is not verified or missing sessionToken then offer to sign in or confirm
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/pair/index.js]: Is it possible for the account to be missing a sessionToken? Does that happen in PB mode?

File: packages/fxa-content-server/app/scripts/views/pair/pairing-totp-mixin.js [hunk 3]
```diff
@@ -6,7 +6,7 @@ import AuthErrors from '../../lib/auth-errors';
 
 export default function () {
   return {
-    checkTotpStatus() {
+    checkTotpStatus(sessionToken) {
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/pair/pairing-totp-mixin.js]: The passing in of the sessionToken but depending on the account from `this.getSignedInAccount()` seems strange to me. How would the sessionToken returned by getSignedInAccount be different to the one passed in?

Is it possible for the browser to return an account with no sessionToken which does not cause the signedInAccount to update it's sessionToken?

File: packages/fxa-content-server/app/scripts/views/pair/index.js [hunk 4]
```diff
@@ -21,22 +22,27 @@ class PairIndexView extends FormView {
...
+      return this.navigateAway(this.getEscapedSyncUrl('signin', 'fxa:pair'));
     }
 
-    return this.checkTotpStatus();
+    // here we pass a token from the 'browserSignedInAccount'. This token has a device
+    // attached to it and is the one we want to check. We want to avoid using the
+    // "web" session token because it may disappear, expire or get pruned.
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/pair/index.js]: I see what's going on now answering the question of why. What might be conceptually simpler is to continue to use the getSignedInAccount that's stored in the user, but to avoid the problems you mention here, update [app-start.js->_getUserStorageInstance](https://github.com/mozilla/fxa/blob/74d4d8e15160dc3524788d2c5c313971ce2d867e/packages/fxa-content-server/app/scripts/lib/app-start.js#L575) to use the memory storage if you are pairing, and update [user.js->shouldSetSignedInAccountFromBrowser](https://github.com/mozilla/fxa/blob/74d4d8e15160dc3524788d2c5c313971ce2d867e/packages/fxa-content-server/app/scripts/models/user.js#L725) if you are pairing.

File: packages/fxa-content-server/app/scripts/views/pair/index.js [hunk 5]
```diff
@@ -21,22 +22,27 @@ class PairIndexView extends FormView {
...
       return this.replaceCurrentPage('pair/unsupported');
     }
 
+    const account = this.broker.get('browserSignedInAccount');
     // If we reach this point that means we are in Firefox Desktop
-    if (! this.broker.get('browserSignedInAccount')) {
-      // if we are not logged into Sync then we offer to sign in
+    if (! account) {
+      // if we are not logged into Sync then we offer to connect
       return this.replaceCurrentPage('connect_another_device');
```
Comment (shane-tomlinson) [inline packages/fxa-content-server/app/scripts/views/pair/index.js]: Is connect another device correct here? Wouldn't we want the user to sign in?

File: packages/fxa-content-server/app/scripts/views/pair/index.js (hunk 2 above)
Comment (vladikoff) [reply packages/fxa-content-server/app/scripts/views/pair/index.js]: not in PB mode, but it happened to somebody in user research :(

Comment (shane-tomlinson) [reply packages/fxa-content-server/app/scripts/views/pair/index.js]: > not in PB mode, but it happened to somebody in user research :(

Amazing... :( seems about right.

File: packages/fxa-content-server/app/scripts/views/pair/index.js (hunk 5 above)
Comment (vladikoff) [reply packages/fxa-content-server/app/scripts/views/pair/index.js]: ah yes! We can now use the new `navigateAway` thing here as well, good catch!

**Comment by vladikoff** [comment]
@shane-tomlinson updated!
